Wrappers/Python/ccpi/optimisation/functions/MixedL21Norm.py

```diff
@@ -19,105 +19,164 @@
 from __future__ import absolute_import
 from __future__ import division
 from __future__ import print_function
-from __future__ import unicode_literals
 
-from ccpi.optimisation.functions import Function, ScaledFunction
+
+from ccpi.optimisation.functions import Function
 from ccpi.framework import BlockDataContainer
 import numpy as np
 
-import functools
 
 class MixedL21Norm(Function):
     
     
-    '''
-        f(x) = ||x||_{2,1} = \sum |x|_{2}                   
-    '''      
+    """ MixedL21Norm function: :math:`F(x) = ||x||_{2,1} = \sum |x|_{2} = \sum \sqrt{ (x^{1})^{2} + (x^{2})^{2} + \dots}`                  
+    
+        where x is a BlockDataContainer, i.e., :math:`x=(x^{1}, x^{2}, \dots)`
+    
+    """      
     
     def __init__(self, **kwargs):
 
-        super(MixedL21Norm, self).__init__()                      
+        super(MixedL21Norm, self).__init__()  
+        self.b = kwargs.get('b', None)  
+
+        # This is to handle tensor for Total Generalised Variation                    
         self.SymTensor = kwargs.get('SymTensor',False)
         
-    def __call__(self, x):
+        # We use this parameter to make MixedL21Norm differentiable
+#        self.epsilon = epsilon
         
-        ''' Evaluates L2,1Norm at point x
+        if self.b is not None and not isinstance(self.b, BlockDataContainer):
+            raise ValueError('__call__ expected BlockDataContainer, got {}'.format(type(self.b)))
             
-            :param: x is a BlockDataContainer
-                                
-        '''
+        
+    def __call__(self, x):
+        
+        r"""Returns the value of the MixedL21Norm function at x.                                            
+        """
         if not isinstance(x, BlockDataContainer):
             raise ValueError('__call__ expected BlockDataContainer, got {}'.format(type(x))) 
-                                         
-        tmp = x.get_item(0) * 0.
-        for el in x.containers:
-            tmp += el.power(2.)
-        return tmp.sqrt().sum()
+            
+#        tmp_cont = x.containers                        
+#        tmp = x.get_item(0) * 0.
+#        for el in tmp_cont:
+#            tmp += el.power(2.)
+#        tmp.add(self.epsilon**2, out = tmp)    
+#        return tmp.sqrt().sum()            
+         
+        y = x
+        if self.b is not None:
+            y = x - self.b    
+        return y.pnorm(p=2).sum()                                
```

Comment:
`pnorm` is to get the euclidean norm of DataContainers, returning a DataContainer and not a number